Schedule rework

mschuwalow opened the pull request:
Sorry for taking forever with this.
This is an initial version of the schedule changes. I'll definitely need to another pass to update docs, double check / cleanup implementations and unfuck Stream#aggregateWithinEither which I currently broke. I plan to spend some more time tomorrow to polish this.

Still I believe this is developed enough to take a look at the new api and decide whether we want to stick with it.
In general I think this is very clean to program with and we lost only very few combinators in the process.
One thing that is a bit unintuitive with the new encoding is building delays in a composable fashion. In order to make this easier I split up all time-based schedules into functions that produce non-sleeping schedules that emit durations and a function addDelay that allows to 'materialize' this into sleeps. This seems to work alright and feels very similar to programming with streams. I don't know whether that is the right approach though. I think @adamgfraser had ideas for other possible encodings.

@jdegoes @adamgfraser It would be great if you could take a look

adamgfraser commented:
@mschuwalow Thanks for taking the lead on this! This is really great work and goes a long way to cleaning up a complicated part of the codebase with a lot of dependencies.

The main concern I have with the current implementation is that it pushes more work onto the user to specify whether a `Schedule` should do "sleeps" or not. For example, consider this snippet:

```scala
ZIO.effectTotal(println("Hello, World!")).repeat(ZSchedule.fixed(1.second))
```

In the current version of ZIO, this would print "Hello, World!" to the console repeatedly with a one second delay. In this PR, that line would instead immediately print "Hello, World!" an infinite number of times with no delay. To get an actual delay we need to instead do:

```scala
ZIO.effectTotal(println("Hello, World!")).repeat(ZSchedule.fixed(1.second).addDelay)
```

And the user in general has to do that every time they call a `Schedule` method. We generally can't do it within ZIO library combinators because `addDelay` adds a clock dependency so is only appropriate if we know that the schedule will always involve "sleeps" and not just for example be repeating while a condition is true. This is not great from a user perspective. Incidentally if we do go down this road I would suggest renaming `addDelay` to something like `run` to highlight its nature of adding real `Clock` effects and being irreversible (we can `addDelay` to any schedule but we can't do that reverse because at that point we have actual sleep effects that we can't introspect).

I think broadly this PR goes in the right direction by continuing to model a schedule as a "description" of a series of steps with delays rather than in each step doing the delay, at least until the schedule is "run" with `addDelay`.  We generally benefit by describing effects versus actually doing them and this true here as well where if each step does a "sleep" in combinators like `spaced` we lose the ability to modify those delay in some way because by the time we "see" them they have already occurred, so for example we can add delays but can't reduce them and `||` would need to behavior more like `ZSteam.merge`. I think there is a consistent set of combinators there but they are different than the existing one and in some ways more restrictive, so I think you are right to model most combinators as descriptions and only introduce actually doing the "sleeping" as late as possible.

What I think we need to add is using the type system more to automatically "derive" whether a schedule should use `Clock` and do sleeping, versus the user having to specify that using `addDelay`. Conceptually, the compiler should "know" that a schedule using `Schedule.fixed` should do real sleeps and that one only using `Schedule.doWhile` shouldn't. To do that, every schedule needs to maintain one additional piece of information, which we could think of as the "ExecutionStrategy" for the schedule, which could either be "live" (do sleeping), or not. Then I think the composition rule is that if we combine schedules and either schedule uses "sleep", the combined combinator uses sleep. I think if we take that aspect from #1720 and incorporate it we could get schedules that automatically propagate the appropriate clock dependency, which I think is ultimately where we want to be.

If that is right then I think we conceptually end up at a similar place to #1720 (schedules are descriptions of a series of delays that include an instruction for how a client can "run" those delays). And then I think we have to see how different they. If schedules really are descriptions of a series of delays with a way for a client to effectually "run" them to potentially actually do a "sleep", then we can implement all downstream method with almost no change and remove the `Clock` dependency as in #1720.

jdegoes commented:
>  In order to make this easier I split up all time-based schedules into functions that produce non-sleeping schedules that emit durations and a function addDelay that allows to 'materialize' this into sleeps. 

I too am not a fan of this. It impacts user-land ergonomics enormously.

What is wrong with just sleeping? Seems like you almost had that working at the hackathon. Run into more problems there???

mschuwalow commented:
> > In order to make this easier I split up all time-based schedules into functions that produce non-sleeping schedules that emit durations and a function addDelay that allows to 'materialize' this into sleeps.
> 
> I too am not a fan of this. It impacts user-land ergonomics enormously.
> 
> What is wrong with just sleeping? Seems like you almost had that working at the hackathon. Run into more problems there???

I guess the one that is specifically proving challenging is jittered.
I believe having the option to at least randomly extend how long schedules are sleeping is super useful, but it's not quite clear to me how to express this on top of the new api. One option is see is
measuring how long schedules are taking with timed and extending it with a second sleep, but this feels quite dirty.
The second is the approach I've taken in the pr, but I agree that the ergonomics are bad.

@jdegoes Do you have an idea how to express this?

adamgfraser commented:
Isn’t the problem more fundamental? Take the example of:

```scala
ZSchedule.spaced(10.seconds).modifyDelay(_ / 2)
```

If a schedule just describes delays, so each step doesn’t actually sleep but returns some state indicating how long to sleep this is easy because we just divide the number by two.

But if a schedule “sleeps” then how do we implement this? We just have an opaque effect that we can either do nothing with or run, in which case we actually sleep for 10 seconds and it is already too late to achieve the goal of sleeping for only 5 seconds. By converting schedules from descriptions to actions we have lost some ability to compose them.

We could instead implement `modifyDelay` as only adding additional delays and `jittered` as increasing each delay by a random amount instead of potentially increasing or reducing the delay, but I think that is just one example of how the semantics of schedules that just sleep will be somewhat different than the existing behavior.

iravid commented:
Just throwing something out here. What if we add a delay "instruction", represented by an ADT that is parameterized by `R`? This ADT can be translated to a ZIO but it keeps the `R`.

Something like ...
```scala
sealed trait Something[-R]
object Something {
  case object DoesNothing extends Something[Any]
  case object AddsDelay(delay: Duration) extends Something[Clock]
}
```
Keep this on the decision, and you can both introspect it and translate it to a ZIO while keeping the R precise. What do you  think @mschuwalow @adamgfraser ?

adamgfraser commented:
@iravid Yes that is exactly the direction I think we need to go in. Here is what I had in #1720 that does this. Obviously we could change the names.

```scala
sealed trait Delay[-R] extends (Duration => ZIO[R, Nothing, Unit]) {
  def combine[R1 <: R](that: Delay[R1]): Delay[R1]
}
final case object Live extends Delay[Clock] {
  def apply(d: Duration): ZIO[Clock, Nothing, Unit] =
    clock.sleep(d)
  def combine[R1 <: Clock](that: Delay[R1]): Delay[R1] =
    this
}
final case object Mock extends Delay[Any] {
  def apply(d: Duration): ZIO[Any, Nothing, Unit] =
    ZIO.unit
  def combine[R1 <: Any](that: Delay[R1]): Delay[R1] =
    that
}
```

regiskuckaertz commented:
@adamgfraser It seems with that encoding, you will always end up writing `decision.delay(decision.duration)`. Why not specialise `Decision` itself?
```scala
sealed abstract class Decision[R, A, B](cont: Boolean, state: A, finish: () => B) {
  def delay: ZIO[R, Nothing, Unit] = ZIO.unit
}
final case class Sleep(duration: Duration, cont: Boolean, state: A, finish: () => B) extends Decision[Clock, A, B](cont, state, finish) {
  def delay = clock.sleep(duration)
}
final case class At(instant: Instant, cont: Boolean, state: A, finish: () => B) extends Decision[Clock, A, B](cont, state, finish) {
  def delay = clock.now.map(instant.toEpochMillis - _).flatMap(n => if (n <= 0) ZIO.unit else clock.sleep(n))
}
```

iravid commented:
@regiskuckaertz nice!!

adamgfraser commented:
@regiskuckaertz Yes you could definitely do that. You end up not having to write it very often because you do it in a few key combinators like `retry` and `repeat` and then from a user perspective it is completely transparent. In my original version I was trying to remove the `Clock` dependency with minimal other changes so it worked well to leave `duration` as a field on `Decision`.

adamgfraser commented:
One advantage of this approach is that it should be possible to implement all the methods built on top of `Schedule` with almost no changes. You should just have to change any references to `clock.sleep(duration)` to `decision.delay`.

mschuwalow commented:
I really like @regiskuckaertz encoding if we want to go with this style. Especially the At decision would be a very nice way to tackle the cron stuff

regiskuckaertz commented:
Also adding a case that cannot make progress anymore would fix `cont`:
```scala
sealed abstract class Decision[R, A, B](cont: Boolean)
Final case class Done[B](b: B) extends Decision[Any, Nothing, B](false)
```
but I’m not sure it would be useful.

On Mon, 23 Sep 2019 at 16:46, Maxim Schuwalow <notifications@github.com>
wrote:

> I really like @regiskuckaertz <https://github.com/regiskuckaertz>
> encoding if we want to go with this style. Especially the At decision would
> be a very nice way to tackle the cron stuff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zio/zio/pull/1798?email_source=notifications&email_token=AAEZZWELC6ZARPY3WMDYWCLQLDQFVA5CNFSM4IZFCJN2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7LKFBI#issuecomment-534160005>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEZZWCX7VDEM3TD42GWFCDQLDQFVANCNFSM4IZFCJNQ>
> .
>
-- 
–Regis

-- 

This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 

Guardian News & Media Limited is a member of Guardian Media Group plc. 
Registered Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP. 
 Registered in England Number 908396

mschuwalow commented:
Hm maybe we can combine the two approaches. What if we keep the encoding in this pr, which allows nice fold-based usage instead of cont, but introduce an `Delayed` abstraction very close to what @regiskuckaertz proposed. Something like:
```scala
trait ZSchedule0[-R, -A, +B] {
  type State
  type RunningState <: State

  val initial: ZIO[R, Nothing, State]
  val extract: RunningState => B
  val update: (A, State) => ZIO[R, RunningState, Delayed[R, RunningState]]
}

object ZSchedule0 {

  sealed abstract class Delayed[-R, +A] {
    def delay: ZIO[R, Nothing, A]
  }
  final case class Sleep[A](duration: Duration, result: A) extends Delayed[Clock, A] {
    val delay = clock.sleep(duration).as(result)
  }
  final case class At[A](instant: Instant, result: A) extends Delayed[Clock, A] {
    val delay = clock.currentTime(TimeUnit.MILLISECONDS).map(instant.toEpochMilli - _).flatMap(n => if (n <= 0) ZIO.unit else clock.sleep(n.millis)).as(result)
  }
}
```
This also makes it much clearer that a schedule is not supposed to delay when it short circuits. We could even make this more polymorphic in order to make this more explicit, at the cost of another parameter:
```scala
trait ZSchedule0[-R0, -R, -A, +B] {
  type State
  type RunningState <: State

  val initial: ZIO[R0, Nothing, State]
  val extract: RunningState => B
  val update: (A, State) => ZIO[R0, RunningState, Delayed[R, RunningState]]
}
```

jdegoes commented:
@mschuwalow 

> I guess the one that is specifically proving challenging is jittered.

If you make `jittered` only work on schedules with `Clock with Random`, it becomes trivial.

In general, `modifyDelay` could work only on `Clock with Random`.

We have seen in #676 that it is not sufficient to have a `Duration`. It's a leaky abstraction because it does not permit you to create schedules that run at specific times. The very concept of `modifyDelay` assumes all delays are relative, not absolute; which is true for some common cases but not other common cases (cron-style).

Essentially, I believe that losing `modifyDelay` (and friends) is a good thing. We throw away some structure (which has cost, of course!) and get something that is more general, which can satisfy more use cases. Yes, we have fewer operations, but then the data type becomes less constrained, which means we can use it for broader use cases.

> @jdegoes Do you have an idea how to express this?

I would press forward with the generic approach. It results in less code, cleaner code, and yes, fewer operations, but a more capable data type. The alternative will be adding even more and highly irregular structure (you can see that emerging in #676). 

For `jitter` in particular, you are right that it is useful. I would make `jitter` work only on `Clock with Random`, or you can easily make it work on `R >: DefaultEnvironment#Environment <: Clock with Random`, i.e. set of all ZIO built-in services, but at least `Clock` and `Random` for the jitter. That takes care of the common useful case in a way that's repeatable (we can do the same thing elsewhere, if we want), and we can then sacrifice `modifyDelay` and such as being too specialized and not compatible with scheduling at absolute date/times.

mschuwalow commented:
Alright, I'll update the pr and get back to you

mschuwalow commented:
True, you would probably end up special casing behavior for the different sleep styles all the time O.o
I've added this formulation of jittered. It's a bit sad that we can't write this fully generically without macros :cry: 

There is one further issue I ran into with this encoding.
We currently are unable to extract the output from a schedule without updating once, which is desired behavior. The problem arises when we want to interrupt the schedule, but still want to be able to extract the output from it. This happens whenever we race two schedules and require both of their outputs.

The prime candidates where this is necessary are `&&` and `***`.
I don't think it's possible to implement intersection without making the type signatures very unergonomic and Arrow's split at all.

I believe the following encoding works for all we want to do, but does not enforce the correct usage as strongly.
```scala
trait ZSchedule[-R, -A, +B] extends Serializable { self =>
  type State
  val initial: ZIO[R, Nothing, State]
  val extract: (A, State) => B
  val update: (A, State) => ZIO[R, Unit, State]
```
Where initial returns the state to be paired with the first element and update($a_1, $state_1) returns state_2 and so on. Alternatively we might be able to do this, but I don't whether this is any better and I haven't tried it out.
```scala
trait ZSchedule[-R, -A, +B] extends Serializable { self =>
  type State
  val initial: A => ZIO[R, Nothing, State]
  val extract: State => B
  val update: (A, State) => ZIO[R, Unit, State]
```
What do you think?

/cc @jdegoes

iravid commented:
I admit that just by having a cursory look, there are quite a few semantic and signature changes than I feel comfortable with.

Are all of them necessitated by the changes required to support a specific `R`? Or can we roll back some of the encoding changes and save them for 2.0?

iravid commented:
In particular I should note that the `initial/update/extract` style of usage, seen in ZSink, is *very* hard to get right across all combinators and constructors.

adamgfraser commented:
We could merge #1720, which would remove the `Clock` dependency with basically no changes to other combinators, and then use this PR as the basis for additional work. That would let us start implementing other combinators on `ZIO` in terms of `Schedule`, which we're already seeing more requests for, and potentially alleviate some of the pressure to get this in immediately.

iravid commented:
@mschuwalow What limitations on #1720 led to exploring a different encoding?

mschuwalow commented:
@iravid @adamgfraser 
I believe the main motivation for this is having a better basis to implement more sophisticated behavior on top of schedules (e.g. #676).
I agree though that this is shaping up to be a rather large change and it might make sense to delay this in favor of #1720. Especially as the design space for this seems quite large and we might need a few more iterations

mschuwalow commented:
Ok, I believe this is at a point where this can be reviewed. I'm still unsure about the specific encoding of schedule, but I believe the current semantics work out nicely for everything we want to do.

jdegoes commented:
I will take a detailed look this PR and the other one this week.

The main goal is allowing schedules that run at absolute times, so we can define compositional cron-like languages; secondary goal would be eliminating `Clock` requirement from `repeat` / `retry` combinators, if they use schedules that do not require `Clock`; third goal would be simplifying the core `Schedule` model, if possible. 

I think the main goal (cron) is worth pushing in a "big enough" change to get this through (it is a long requested feature); ideally the reworking would result in something not harder to use than the old API.

mschuwalow commented:
@jdegoes did you have an opportunity take a look yet?

jdegoes commented:
@mschuwalow @adamgfraser 

Here is my analysis of #1798: 

 - [x] Support schedules that run at absolute times
 - [x] Eliminate `Clock` requirement from repeat/retry combinators
 - [x] Simplify core `Schedule` model

Here is my analysis of #1720: 

 - [ ] Support schedules that run at absolute times
 - [x] Eliminate `Clock` requirement from repeat/retry combinators
 - [ ] Simplify core `Schedule` model

I further note that in #1798, we managed to accidentally fix #1909 — in fact, you can see the type signature of `update` demands the correct behavior. It's "fixed in the types".

With @mschuwalow's proposed formulation:

```scala
trait ZSchedule[-R, -A, +B] extends Serializable { self =>
  type State
  val initial: ZIO[R, Nothing, State]
  val extract: (A, State) => B
  val update: (A, State) => ZIO[R, Unit, State]
}
```

...one can also see that `ZSchedule` is a kind of `ZSink`, suggesting a possible future unification and reduction in the core types exposed in ZIO.

On the other hand, compared with #1720, we have several drawbacks:

 - `jittered` and other functions that modify delays need to have fixed `R`; using them with a "bigger" `R` would require delegation in user-land code
 - Because it is more abstract, it is not obvious looking at the schedule code that it has anything to do with delays / intervals
 - Some code that needs to "know" about delays may become more complex and / or rigid, if some other way of solving the problem is not identified

This is of course very subjective, but overall I lean toward #1798 with the proposed encoding; I'd go for that in a second if I saw a prototype `ZSink` encoding for schedule, because it would open up using schedules in every place where a sink is supported.

What are other folks thoughts on this???

regiskuckaertz commented:
It looks like `ZSchedule` and `ZSink` are two encoding of Moore machines:

- the finite set of states is `State`
- the start state is given by `initial`
- the input alphabet is `A`
- the output alphabet is `B`
- the transition function is `update`
- and the output function is `extract`

imho the work "schedule" implies that time is involved in some way, but `ZSchedule` has many combinators that do not wait. What if we moved those to `ZSink`, which would be the "unrestricted" encoding of a Moore machine in ZIO, and `ZSchedule` would be a subtype with the refinement that it is expecting an `R` with a `Clock`? I don't know that it would solve all those encoding issues though.

mschuwalow commented:
> It looks like `ZSchedule` and `ZSink` are two encodings of Moore machines:
> 
> * the finite set of states is `State`
> * the start state is given by `initial`
> * the input alphabet is `A`
> * the output alphabet is `B`
> * the transition function is `update`
> * and the output function is `extract`
> 
> imho the word "schedule" implies that time is involved in some way, but `ZSchedule` has many combinators that do not wait. What if we moved those to `ZSink`, which would be the "unrestricted" encoding of a Moore machine in ZIO, and `ZSchedule` would be a subtype with the refinement that it is expecting an `R` with a `Clock`? I don't know that it would solve all those encoding issues though.

Very cool, I haven't thought about it from this perspective.

With the new encoding schedules are actually Mealy machines and not Moore machines. meaning that they require inputs for both updating and extracting. In practice the difference is just that schedules must have seen at least one input before extracting. The reasoning behind that was that otherwise combinators such as `+++` end up having very wonky semantics if you attempt to extract immediately.

It would be interesting to see if ZSink profits from such an encoding as well.
Other than that you should be able to transform any Moore into a Mealy machine, meaning that if we want to unify them, one can probably have ZSink become a subtype of ZSchedule.

adamgfraser commented:
I'm a little worried about the direction we're going. It feels like we are losing a lot of power to compose schedules because combinators are running delays instead of describing delays. For example, in the old encoding we could take a `Schedule.spaced(10.seconds)` and use `modifyDelay` to cut the delay in half. As I understand it we can't do that in this version because every step in at least that schedule does an actual `sleep`, so we can always sleep longer but we can never sleep less.

#1720 was meant to be a patch fix to remove the `Clock` dependency in the most minimal way possible. It definitely doesn't address the other design goals of supporting scheduling effects at specific times or cleaning up the overall encoding, though I also don't think it makes it any harder to do those things on top of that versus the status quo. What I think we can take away from #1720 is the problem of removing the `Clock` dependency is actually relatively easy to solve.

So if our primary challenge is supporting scheduling actions at a specific time, I wonder if we are better off starting with the concept of a schedule as a description and just adding another constructor, so maybe you have `Decision.Delay` to describe delaying for a finite amount of time, potentially zero, and then `Decision.At` to describe delaying until a specific point in time, and you build an algebra based on that. That gives us the ability to support scheduling actions at a specific point in time and preserves all of the ability to compose schedules that we have today.

Then we can always solve the `Clock` problem by having the schedule include an effect that clients can choose to run to do actual delays if they want to as in #1720 or some of the variants discussed above.

adamgfraser commented:
The discussion of `ZSink` is very interesting. One challenge is that `ZSink` is part of `streams` so we can't use it in `core` to implement a lot of the operations we would want to that currently involve `Schedule` but not `Clock` like `repeat` or `doWhile`. Obviously we could just move it but I think that reflects a deeper reality that right now `ZSink` is built largely around the needs of streams in its use of `Chunk` and concern with issues like handling leftovers which wouldn't be relevant with regard to `ZIO` where every effect produces a single value. Maybe there is a more general version of a sink that belongs in core and then the current sink is a specialization of that? I may be getting out of scope for this or this may be more of a 2.0 design discussion.

jdegoes commented:
> I'm a little worried about the direction we're going.

It's definitely possible. However, if you look at the other PR, you can see what happens when you try to describe both absolute and relative times: comparing them becomes impossible (yet essential because of the semantics of `&&` and `||`). 

So you need an ADT to describe comparison, essentially a min-language for forming predicates on intervals; this leads to other types of complications (infinite thunks, needing to perform effects to see exactly "when" something will run), etc., whichyou can see in the pull request.

When defining `||` with `race` and `&&` with `zipPar`, suddenly the correct semantics become obvious and natural, and independent of absolute time scheduling or relative time scheduling; that hints very strongly we can "forget" about time comparisons and deal at a higher level, one which definitely will forego some capabilities but will also cover more use cases (that's the abstraction tradeoff).

> and use modifyDelay to cut the delay in half. As I understand it we can't do that in this version 

We can do that, although I am not sure yet if it's implemented in this pull request. Essentially, `modifyDelay` and such can be used on any `ZSchedule` for a known environment (e.g. `ZEnv`).

If you believe users won't create schedules that are not composed from the ones in ZIO, then this covers 100% of the use cases. If you believe users will create some custom schedules by using `new ZSchedule`, which use custom `R` (directly or compositionally), then it doesn't solve 100% of use cases, because the user could have a custom `R` for which `modifyDelay` is unsupported, and be unable to push that `modifyDelay` deeper into the composite schedule.

For fixed `R`, we can support exactly the same API as before, with the same semantics. I don't know for sure, but I believe that will cover a very high percentage of use cases (probably all of the ones in the wild), without the complications that come from modeling a delay as absolute or relative time (which seems simple but becomes quite complex very fast and breaks many APIs).

jdegoes commented:
> Maybe there is a more general version of a sink that belongs in core and then the current sink is a specialization of that? I may be getting out of scope for this or this may be more of a 2.0 design discussion.

I agree on both accounts. I think a mealy machine is lurking here and the stream one is a specialization of that one, and should probably be encoded as such; and the schedule may be a different specialization.

mschuwalow commented:
> We can do that, although I am not sure yet if it's implemented in this pull request. Essentially, `modifyDelay` and such can be used on any `ZSchedule` for a known environment (e.g. `ZEnv`).

`modifyDelay` is currently one of the few combinators not ported, mostly because I forgot about it. I'll include that one when I update the pr the next time.

The approach I have taken with jittered, and what I would also do with modifyDelay is having a very flexible definition where https://github.com/zio/zio/blob/85a1928db582c432b99752b4d1df79c123fd5b0d/core/shared/src/main/scala/zio/ZSchedule.scala#L352-#L379 the 'modified' environment has to be manually spliced into the old environment, and having specialization for the default environment of java and js respectively.
https://github.com/zio/zio/blob/85a1928db582c432b99752b4d1df79c123fd5b0d/core/jvm/src/main/scala/zio/ZScheduleSyntax.scala.

I think this is a nice tradeoff where the common cases are very comfortable, but we can still support custom environments with minimal boilerplate.

mschuwalow commented:
> I'm a little worried about the direction we're going. It feels like we are losing a lot of power to compose schedules because combinators are running delays instead of describing delays. For example, in the old encoding we could take a `Schedule.spaced(10.seconds)` and use `modifyDelay` to cut the delay in half. As I understand it we can't do that in this version because every step in at least that schedule does an actual `sleep`, so we can always sleep longer but we can never sleep less.

I was worried about that as well, but I think in practice usage ends up being quite nice. All combinators that are incremental the way that are building effects and delays are very straightforward and I don't think we are losing anything there.

I think the formulation in this pr does have a disadvantage for combinators that modify existing delays or effects. Here we have to rely on interacting with the services directly, as is currently done for `jittered` and could also be done for `modifyDelay` in the same way.

mschuwalow commented:
> The discussion of `ZSink` is very interesting. One challenge is that `ZSink` is part of `streams` so we can't use it in `core` to implement a lot of the operations we would want to that currently involve `Schedule` but not `Clock` like `repeat` or `doWhile`. Obviously we could just move it but I think that reflects a deeper reality that right now `ZSink` is built largely around the needs of streams in its use of `Chunk` and concern with issues like handling leftovers which wouldn't be relevant with regard to `ZIO` where every effect produces a single value. Maybe there is a more general version of a sink that belongs in core and then the current sink is a specialization of that? I may be getting out of scope for this or this may be more of a 2.0 design discussion.

At least with the encoding / semantics in this pr for schedule and the current one for sinks the main difference between sinks and schedules is that schedules really want to have seen at least one input before extracting. Actually the schedules current encoding is basically a direct implementation of mealy machines https://en.wikipedia.org/wiki/Mealy_machine with an effectful update function.

On the other hand, as @regiskuckaertz pointed out, sinks can be encoded as moore machines. https://en.wikipedia.org/wiki/Moore_machine. An interesting difference here is that sink currently has an effectful extract function as well.

At least from my understanding moore machines (sinks) can be quite easily used as mealy machines (schedules) while the inverse is not true in general. Still unifying them could be very interesting, especially considering how they relate conceptually

jdegoes commented:
> I think the formulation in this pr does have a disadvantage for combinators that modify existing delays or effects. Here we have to rely on interacting with the services directly, as is currently done for jittered and could also be done for modifyDelay in the same way.

I think instead of syntax we should put this into the core trait, which just requires something like `ZEnv.ClockProxy`. I'll write that up in a separate ticket.

jdegoes commented:
@mschuwalow 

[Here](https://github.com/zio/zio/issues/1953) is the ticket:

<https://github.com/zio/zio/issues/1953>

> At least with the encoding / semantics in this pr for schedule and the current one for sinks the main difference between sinks and schedules is that schedules really want to have seen at least one input before extracting.

Right, otherwise you can't do `identity` (i.e. it won't form a `Category`, among other things). The sink encoding is a relaxation of this requirement. It would be useful to explore a way of abstracting over that difference without using higher-kinded types.

jdegoes commented:
> modifyDelay is currently one of the few combinators not ported, mostly because I forgot about it. I'll include that one when I update the pr the next time.

I think we should include that and any other "delay modifying ones" (obviously `jitter`) to prove the same things are possible (with fixed `R`) and just as easy as before, API-wise.

mschuwalow commented:
> @mschuwalow
> 
> [Here](https://github.com/zio/zio/issues/1953) is the ticket:
> 
> #1953
> 
> > At least with the encoding / semantics in this pr for schedule and the current one for sinks the main difference between sinks and schedules is that schedules really want to have seen at least one input before extracting.
> 
> Right, otherwise you can't do `identity` (i.e. it won't form a `Category`, among other things). The sink encoding is a relaxation of this requirement. It would be useful to explore a way of abstracting over that difference without using higher-kinded types.

Oh, I really like that solution. I think it makes sense to keep the manual versions around as well for custom environments, but this will be super clean

mschuwalow commented:
> > modifyDelay is currently one of the few combinators not ported, mostly because I forgot about it. I'll include that one when I update the pr the next time.
> 
> I think we should include that and any other "delay modifying ones" (obviously `jitter`) to prove the same things are possible (with fixed `R`) and just as easy as before, API-wise.

Will do!

jdegoes commented on core-tests/shared/src/test/scala/zio/ZScheduleSpec.scala:
```diff
@@ -148,50 +149,61 @@ object ZScheduleSpec
-            val expected                              = List(1, 2, 3, 4, 5).map((1500.millis, _))
-            assertM(scheduled, equalTo(expected))
+            val schedule = ZSchedule.spaced(500.millis).jittered_[Clock, Clock](2, 4) { f => old =>
+              new Clock { val clock = f(old.clock) }
```
These use the more painful `jittered_`

mschuwalow replied:
I'll update the tests to use the new fiberTime that we have in TestClock. This will get cleaned up as part of it :+1:

jdegoes commented on core-tests/shared/src/test/scala/zio/ZScheduleSpec.scala:
```diff
@@ -148,50 +149,61 @@ object ZScheduleSpec
-            val expected                              = List(1, 2, 3, 4, 5).map((250.millis, _))
-            assertM(scheduled, equalTo(expected))
+            val schedule = ZSchedule.spaced(500.millis).jittered_[Clock, Clock](0, 1) { f => old =>
+              new Clock { val clock = f(old.clock) }
```
Ditto

jdegoes commented on core-tests/shared/src/test/scala/zio/ZScheduleSpec.scala:
```diff
@@ -294,6 +307,36 @@ object ZScheduleSpecUtil {
+  /**
+   * Run a schedule using the provided input and collect all outputs
+   */
+  def run[A, B](
```
It seems like we should 't need this amount of ceremony anymore after TestClock, TestRandom etc.

mschuwalow replied:
Yep, this can be trivialized with fiberTime. Working on it as we speak :grin:

iravid commented:
* I looked at the current encoding and I like how it turned out. For a while now the similarities between Sink and Schedule seemed pretty apparent to me - this encoding makes them even more stark!
* as to the similarity with Sink, I’d be happy to move sink out to the core package if that makes sense. For a while now I’ve had a feeling that Sink is a general way to compose aggregations rather than something specifically fused to Stream. Of course if we can abstract over both Sink and Schedule, that abstraction can be in core and Sink’s combinators and constructors could stay in streams.
* I’ll review the changes to the streams combinators thoroughly this week.

Great work @mschuwalow :-)

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

mschuwalow commented:
> * I looked at the current encoding and I like how it turned out. For a while now the similarities between Sink and Schedule seemed pretty apparent to me - this encoding makes them even more stark!
> * as to the similarity with Sink, I’d be happy to move sink out to the core package if that makes sense. For a while now I’ve had a feeling that Sink is a general way to compose aggregations rather than something specifically fused to Stream. Of course if we can abstract over both Sink and Schedule, that abstraction can be in core and Sink’s combinators and constructors could stay in streams.
> * I’ll review the changes to the streams combinators thoroughly this week.
> 
> Great work @mschuwalow :-)
> 
> Sent with [GitHawk](http://githawk.com)

Thanks :)

I guess I will experiment with Sink once this pr is merged. I would really appreciate your feedback then.

Please do review thoroughly, I definitely also need to do another pass and double check all the changes as we have experimented a lot here

jdegoes commented on core/shared/src/main/scala/zio/Schedule.scala:
```diff
@@ -25,22 +25,10 @@ object Schedule {
    * See [[ZSchedule.never]]
    */
-  final val never: Schedule[Any, Nothing] =
+  final val never: Schedule[Any, Unit] =
```
`Unit` doesn't have the same composition properties as `Nothing` — the old `never` could be `>>>` into a `Schedule` requiring any particular type.

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -713,14 +781,25 @@ object ZSchedule {
+   * A schedule that waits forever when updating.
+   */
+  final val never: Schedule[Any, Unit] =
+    ZSchedule[Any, Nothing, Any, Unit](UIO.never, (_, _) => UIO.never, (_, _) => ())
```
Return the `Nothing` that you get from the state, then you can make this `never` succeed with `Nothing` instead of `Unit`

adamgfraser commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -339,19 +345,30 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
+   * Applies random jitter to all sleeps executed by the schedule.
    */
-  final def jittered: ZSchedule[R with Random, A, B] = jittered(0.0, 1.0)
+  final def jittered(min: Double = 0.0, max: Double = 1.0)(implicit ev: ZEnv <:< R): ZSchedule[ZEnv, A, B] =
```
Why not just increase the delay by a random amount? My understanding is that the main use case of `jittered` is to make sure that that we are not inadvertently having a bunch of effects occur at the exact same time and, for example, overwhelming a server. This could be achieved by just adding an additional random delay versus modifying the delay. This would provide a nicer API because we would not have to get into any of these issues about mapping the environment or needing the environment to be a particular type. We could always provide another method that can potentially decrease the delay for users who need that.

I think we should be able to get to the point where for common use cases the environment type can be fully generic and doesn't require special modification and then modifying the environment can be a feature for power users to achieve functionality not available through a fully generic environment type.

jdegoes replied:
That's probably a good point for `jittered` in particular. Of course we still need things like `modifyDelay`, etc.

adamgfraser replied:
Yes exactly.

mschuwalow replied:
I have added ZSchedule#randomDelay and ZSchedule#randomDelayNormal that create a schedule that sleeps with a uniformly and normally distributed delay respectively. Using >>> this can be used to add random delays to arbitrary schedules.

I think jittered is also worth keeping around the way it currently exists.

mschuwalow commented:
I believe this is ready for another round of reviews.
Would be good to collect the remaining open issues so we can get this merged.

iravid commented on streams/shared/src/main/scala/zio/stream/ZStream.scala:
```diff
@@ -380,87 +381,95 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
-      for {
-        decision <- schedule.update(unfoldState.lastBatch, unfoldState.scheduleState)
-        result <- if (!decision.cont)
-                   // When the schedule signals completion, we emit its result into the
```
Can you move this comment to line 461?

iravid commented on streams/shared/src/main/scala/zio/stream/ZStream.scala:
```diff
@@ -380,87 +381,95 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
+        .option
+        .raceEither(unfoldState.nextBatchCompleted.await)
+        .flatMap {
+          case Left(None) =>
```
Would be great if you could comment here on each case who won the race

mschuwalow commented:
> The changes look good to me and I definitely like the new encoding! `update(...).foldM` is much nicer than the previous pattern of update, check the cont flag, delay manually...
> 
> Great work @mschuwalow!

Thanks for reviewing! I'll address your remarks tomorrow and also resolve conflicts with master

reibitto commented:
I checked out this branch and can confirm it solves the particular issue I described in #1909. Thanks @mschuwalow! Looking forward to this.

jdegoes commented on core-tests/shared/src/test/scala/zio/ZScheduleSpec.scala:
```diff
@@ -234,7 +254,7 @@ object ZScheduleSpec
               i   <- alwaysFail(ref).retryOrElseEither(Schedule.once, ioFail)
             } yield i)
-              .foldM(
+              .foldM[Any, Int, String](
```
Are all of these additional type annotations really necessary?

mschuwalow replied:
I double checked and most of them are necessary because otherwise the inferred any will cause warnings

jdegoes commented on core-tests/shared/src/test/scala/zio/ZScheduleSpec.scala:
```diff
@@ -216,7 +236,7 @@ object ZScheduleSpec
               i   <- alwaysFail(ref).retryOrElse(Schedule.once, ioFail)
             } yield i)
-              .foldM(
+              .foldM[Any, Int, String](
```
re all of these additional type annotations really necessary?

jdegoes commented on core/shared/src/main/scala/zio/Schedule.scala:
```diff
@@ -82,7 +71,7 @@ object Schedule {
    * See [[ZSchedule.delayed]]
    */
-  final def delayed[A](s: Schedule[A, Duration]): Schedule[A, Duration] =
+  final def delayed[A](s: Schedule[A, Duration]): ZSchedule[zio.clock.Clock, A, Duration] =
```
No schedule can go into `Schedule` companion object unless `R = Any`. Otherwise you could use `Schedule.XYZ` to create a `ZSchedule`.

If it turns out most of these / many of these can no longer be placed here, we might just rename `ZSchedule` => `Schedule` and put everything in one location...

mschuwalow replied:
Oh, I missed that one.
Yeah I believe removing the Schedule/ZSchedule split makes sense here. I'll do it in a seperate follow up pr

jdegoes commented on core/shared/src/main/scala/zio/ZIO.scala:
```diff
@@ -1028,7 +1028,7 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
    * `io.repeat(Schedule.once)` means "execute io and in case of success repeat `io` once".
    */
-  final def repeat[R1 <: R, B](schedule: ZSchedule[R1, A, B]): ZIO[R1 with Clock, E, B] =
+  final def repeat[R1 <: R, B](schedule: ZSchedule[R1, A, B]): ZIO[R1, E, B] =
```
Awesome to see this, truly awesome! 👍

jdegoes commented on core/shared/src/main/scala/zio/ZIO.scala:
```diff
@@ -1039,7 +1039,7 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
     schedule: ZSchedule[R1, A, B],
     orElse: (E, Option[B]) => ZIO[R1, E2, B]
-  ): ZIO[R1 with Clock, E2, B] =
+  ): ZIO[R1, E2, B] =
```
Ditto

jdegoes commented on core/shared/src/main/scala/zio/ZIO.scala:
```diff
@@ -1049,19 +1049,23 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
-
-    schedule.initial.flatMap(loop(None, _))
+    orElse: (E, Option[B]) => ZIO[R1, E2, C]
+  ): ZIO[R1, E2, Either[C, B]] = {
```
👍

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -52,17 +52,27 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
    * Updates the schedule based on a new input and the current state.
    */
-  val update: (A, State) => ZIO[R, Nothing, ZSchedule.Decision[State, B]]
+  val update: (A, State) => ZIO[R, Unit, State]
```
Way cleaner than before, really quite nice!

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -52,17 +52,27 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
-    combineWith(that)(_ && _, _ max _)
+    new ZSchedule[R1, A1, (B, C)] {
+      type State = (self.State, that.State)
+      val initial = self.initial.zip(that.initial)
```
I think we want to `zipPar` the initial states, too.

mschuwalow replied:
👍

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -71,8 +81,8 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
-      val update = (a: (A, C), s: State) =>
-        self.update(a._1, s._1).zipWith(that.update(a._2, s._2))(_.combineWith(_)(_ && _, _ max _))
+      val extract = (a: (A, C), s: (self.State, that.State)) => (self.extract(a._1, s._1), that.extract(a._2, s._2))
+      val update  = (a: (A, C), s: (self.State, that.State)) => self.update(a._1, s._1).zipPar(that.update(a._2, s._2))
```
So much simpler & more essential, both of these.

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -114,43 +127,44 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
-   * by piping the output of this one into the input of the other, and summing
-   * delays produced by both.
+   * by piping the output of this one into the input of the other. Delays produced by this
+   * schedule will always run before the second schedule is evaluated.
```
Maybe more generally, `Effects described by this schedule will always be executed before the effects described by the second schedule.`

mschuwalow replied:
Much clearer 👍

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -114,43 +127,44 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
-    combineWith(that)(_ || _, _ min _)
+    new ZSchedule[R1, A1, (B, C)] {
+      type State = (self.State, that.State)
+      val initial = self.initial zip that.initial
```
Nice and clean!

We may want to consider a test for [this one](https://github.com/zio/zio/issues/1909) so we can close it out in this PR.

mschuwalow replied:
Good point! added

mschuwalow replied:
I'll also update how the tests are grouped/organized in a seperate pr

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -164,22 +178,16 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
-                  step  <- that.update(a, state)
-                } yield step.bimap(Right(_), Right(_))
-            }
+            self.update(a, v).map(Left(_)) orElse that.initial.flatMap(that.update(a, _)).map(Right(_))
```
Way simpler.

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -248,15 +243,29 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
-   * applied to each delay produced by this schedule.
+   * Returns a new schedule with the given delay added to every update.
+   */
+  final def delayed(f: B => Duration): ZSchedule[R with Clock, A, B] =
```
Hmm, maybe call this `addDelay`, and preserve the old `delayed`? In theory we can still have the old one for `R = ZEnv`.

mschuwalow replied:
👍 Will do

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -248,15 +243,29 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
    */
-  final def delayed(f: Duration => Duration): ZSchedule[R, A, B] =
-    modifyDelay((_, d) => IO.succeed(f(d)))
+  final def delayedM[R1 <: R](f: B => ZIO[R1, Nothing, Duration]): ZSchedule[R1 with Clock, A, B] =
```
Maybe call this `addDelayM`?

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -339,19 +345,30 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
    */
-  final def jittered(min: Double, max: Double): ZSchedule[R with Random, A, B] =
-    modifyDelay((_, d) => nextDouble.map(random => d * min * (1 - random) + d * max * random))
+  final def jittered_[R1](min: Double, max: Double)(
```
I don't feel this use of `_` is consistent with other usages, but I may be wrong. Maybe `jitteredEnv` or `jitteredWithEnv` or something?

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -378,82 +395,108 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
-   * applied to each delay produced by this schedule.
+   * applied to each sleep performed by this schedule.
+   *
+   * Note that this does not apply to sleeps performed in ZSchedule#initial.
```
Is there a reason to exclude that? For backward compatibility?

mschuwalow replied:
We need the current B to decide how to modify sleeps, but sadly cant extract it for initial

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -378,82 +395,108 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
+   * All effects executed while calculating the modified duration will run with the old
+   * environment.
+   */
+  final def modifyDelay_[R1](
```
`modifyDelayEnv`, `modifyDelayWithEnv`?

mschuwalow replied:
I like modifyDelayEnv. Will update 👍

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -378,82 +395,108 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
-   * Returns a new schedule that reconsiders the decision made by this schedule.
+   * Provide all requirements to the schedule.
    */
-  final def reconsider[A1 <: A, C](
```
It would be a shame to lose `reconsider` / `reconsiderM`, especially since we can keep them with `R = ZEnv`.

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -378,82 +395,108 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
-    f: (A1, ZSchedule.Decision[State, B]) => ZSchedule.Decision[State, C]
-  ): ZSchedule[R, A1, C] =
-    reconsiderM((a, s) => IO.succeed(f(a, s)))
+  final def provide(r: R): ZSchedule[Any, A, B] =
```
❤️

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -378,82 +395,108 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
+  /**
+   * Provide some of the requirements to the schedule.
+   */
+  final def provideSome[R1](f: R1 => R): ZSchedule[R1, A, B] =
```
👍

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -378,82 +395,108 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
-    f: (A1, ZSchedule.Decision[State, B]) => UIO[ZSchedule.Decision[State, C]]
-  ): ZSchedule[R, A1, C] =
+  final def reconsider[R1 <: R, A1 <: A](
+    f: (A1, Either[State, State]) => ZIO[R1, Unit, State]
```
This isn't quite equivalent to the old reconsider, nor can it be without fixing the env.

mschuwalow replied:
I'm not quite sure what we are losing compared to the old reconsider.
Can you explain in a little more detail?

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -615,11 +655,10 @@ object ZSchedule {
    */
-  final def delayed[R, A](s: ZSchedule[R, A, Duration]): ZSchedule[R, A, Duration] =
-    s.modifyDelay((b, d) => IO.succeed(b + d)).reconsider((_, step) => step.copy(finish = () => step.delay))
+  final def delayed[R, A](s: ZSchedule[R, A, Duration]): ZSchedule[R with Clock, A, Duration] =
```
Better way: `def delayed[R <: Clock, ...]` It will infer better!

mschuwalow replied:
👍

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -662,29 +701,88 @@ object ZSchedule {
-    delayed(unfold[(Duration, Duration)]((Duration.Zero, one)) {
-      case (a1, a2) => (a2, a1 + a2)
-    }.map(_._1))
+  final def fibonacci(one: Duration): ZSchedule[Clock, Any, Duration] =
```
I really like how the `R` tells you if the schedule sleeps or not now. More information in the type.

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -713,14 +811,60 @@ object ZSchedule {
    * that in `(op: IO[E, A]).repeat(Schedule.recurs(0)) `, op is not done at all.
    */
-  final def recurs(n: Int): Schedule[Any, Int] = forever.whileOutput(_ <= n)
+  final def recurs(n: Int): Schedule[Any, Int] = forever.whileOutput(_ < n)
```
Intentional change?

mschuwalow replied:
Sadly check subtly changed its semantics as we now have to extract before updating and not after.
This is a consequence of that change

jdegoes commented on test-tests/shared/src/main/scala/zio/test/TestUtils.scala:
```diff
@@ -41,7 +40,7 @@ object TestUtils {
       .handle { case _ => (false, fail(label)) }
 
-  final def nonFlaky[R, E](test: ZIO[R, E, Boolean]): ZIO[R with Clock, E, Boolean] =
+  final def nonFlaky[R, E](test: ZIO[R, E, Boolean]): ZIO[R, E, Boolean] =
```
❤️ ❤️ ❤️

mschuwalow commented:
@jdegoes thanks for the review and the kind feedback 🙏 
I believe I have addressed most remarks in your review, but there were a few question that arose. I have left the relevant comments unresolved. Would be great if you could take another look -- I'll finish the changes asap after.

jdegoes commented on core/shared/src/main/scala/zio/ZSchedule.scala:
```diff
@@ -281,12 +354,18 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
+    updated(
+      update =>
+        (a: A, s: State) =>
+          update(a, s).foldM(
```
`update(a, s) tapError finalizer`
